### ISSUE TITLE
[BUG] Problem when use "--cross-entropy-fusion-impl te"

### ISSUE THREAD

**Issue by HyunHo99** [NVIDIA/Megatron-LM#1627]
**Describe the bug**
When running the MoE model CPT, using the "--cross-entropy-fusion-impl te" option causes the loss to diverge and the Grad_norm to behave abnormally smoothly. Removing this option resolved the issue.

**To Reproduce**
I used this Megatron option to train the MoE model.

```shell
options=" \
    --tensor-model-parallel-size 1 \
    --pipeline-model-parallel-size 2 \
    --context-parallel-size 1 \
    --expert-tensor-parallel-size 1 \
    --expert-model-parallel-size 8 \
    --use-distributed-optimizer \
    \
    --num-layers 48 \
    --hidden-size 2048 \
    --ffn-hidden-size 6144 \
    --num-attention-heads 32 \
    --num-query-groups 4 \
    --kv-channels 128 \
    --seq-length $SEQ_LEN \
    --max-position-embeddings $SEQ_LEN \
    --use-rotary-position-embeddings \
    --position-embedding-type rope \
    --rotary-base $ROPE_BASE \
    \
    --micro-batch-size 1 \
    --global-batch-size 8192 \
    --train-samples $TRAIN_SAMPLES \
    --lr-warmup-fraction 0.1 \
    --lr 6e-5 \
    --min-lr 6e-6 \
    --lr-decay-style cosine \
    --lr-decay-samples $LR_DECAY_SAMPLES \
    \
    --log-interval 10 \
    --eval-iters 10 \
    --eval-interval 1000 \
    --reset-position-ids \
    --no-create-attention-mask-in-dataloader \
    \
    --data-path $TRAIN_DATASET_PATH \
    --tokenizer-type HuggingFaceTokenizer \
    --tokenizer-model $TOK_PATH \
    --data-cache-path $IDX_PATH \
    --load $BASE_CHECKPOINT_PATH \
    --no-load-optim \
    --no-load-rng \
    --save $CHECKPOINT_PATH \
    --save-interval 5000 \
    --split 99,1,0 \
    --clip-grad 1.0 \
    --weight-decay 0.1 \
    --adam-beta2 0.95 \
    --init-method-std 0.01 \
    \
    --num-experts 128 \
    --moe-layer-freq '([1]*48)' \
    --moe-ffn-hidden-size 768 \
    --moe-router-load-balancing-type aux_loss \
    --moe-router-pre-softmax \
    --moe-router-topk 8 \
    --moe-token-dispatcher-type alltoall \
    --moe-aux-loss-coeff 0.001 \
    --moe-grouped-gemm \
    --moe-permute-fusion \
    \
    --tensorboard-dir $TENSORBOARD_PATH \
    --log-validation-ppl-to-tensorboard \
    --bf16 \
    --sequence-parallel \
    --use-flash-attn \
    --normalization RMSNorm \
    --qk-layernorm \
    --swiglu \
    --untie-embeddings-and-output-weights \
    --no-load-optim \
    --no-load-rng \
    --group-query-attention \
    \
    --cross-entropy-loss-fusion \
    --cross-entropy-fusion-impl te \
    --recompute-activations \
    \
    --overlap-grad-reduce \
    --overlap-param-gather \
    --timing-log-level 2 \
    --no-one-logger \
    --log-throughput \
    --disable-bias-linear \
    --norm-epsilon 1e-6 \
    --adam-beta1 0.9 \
    --attention-dropout 0.0 \
    --hidden-dropout 0.0 \
    "
``` 
**Expected behavior**
<img width="452" alt="Image" src="https://github.com/user-attachments/assets/4edcd22b-6305-423d-b101-a98aa51cf240" />
<img width="467" alt="Image" src="https://github.com/user-attachments/assets/82ec6a43-eebc-470a-bde7-42fa078b79a4" />
The pink line represents when the "te" option is enabled, while the orange line represents when the option is disabled.

When the option is enabled, the grad norm becomes abnormally smooth, and the loss graph shows unusual tendencies.

**Environment (please complete the following information):**
 - Megatron-LM commit ID: 3235d439c9d76b62cf34108abb366013b4833cbf
 - PyTorch version: I used "nvidia/pytorch:25.04-py3" docker image
 - CUDA version: 12.9
 - NCCL version: 2.26.3+cuda12.9

**Comment by Victarry** [comment]
Thanks for reporting. We are currently investigating this issue. + @BestJuly

**Comment by BestJuly** [comment]
Hi @HyunHo99 , could you try the [PR](https://github.com/NVIDIA/TransformerEngine/pull/1879) in TE? It should fix this issue.

**Comment by HyunHo99** [comment]
By applying the modifications, the issue has been resolved! Thank you for addressing it so quickly.

<img width="398" alt="Image" src="https://github.com/user-attachments/assets/c1ea15b6-e7de-4ccf-95d2-408ce1deddb7" />

**Comment by casper-hansen** [comment]
I am using TE 2.5.0 and it's not fixed - see https://github.com/NVIDIA/Megatron-LM/issues/1750#issuecomment-3237705895

<img width="727" height="313" alt="Image" src="https://github.com/user-attachments/assets/4906a258-161d-4116-8f58-a74151d3ee14" />

<img width="722" height="309" alt="Image" src="https://github.com/user-attachments/assets/0ada742e-d19c-4690-80f2-4b46afbeb320" />
